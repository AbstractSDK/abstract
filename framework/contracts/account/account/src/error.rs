use cosmwasm_std::StdError;

use manager::error::ManagerError;
use proxy::error::ProxyError;
use thiserror::Error;

#[derive(Error, Debug, PartialEq)]
pub enum AccountError {
    #[error("{0}")]
    Std(#[from] StdError),

    // #[error("{0}")]
    // Abstract(#[from] AbstractError),

    // #[error("{0}")]
    // AbstractSdk(#[from] AbstractSdkError),

    // #[error("{0}")]
    // Admin(#[from] AdminError),
    #[error(transparent)]
    Manager(#[from] ManagerError),

    #[error(transparent)]
    Proxy(#[from] ProxyError),
}
