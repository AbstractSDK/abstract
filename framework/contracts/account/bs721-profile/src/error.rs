use cosmwasm_std::StdError;
use cw_controllers::AdminError;
use cw_ownable::OwnershipError;
use cw_utils::PaymentError;
use thiserror::Error;

#[derive(Error, Debug)]
pub enum ContractError {
    #[error("{0}")]
    Std(#[from] StdError),

    #[error("{0}")]
    OwnershipError(#[from] OwnershipError),
    #[error("{0}")]
    Payment(#[from] PaymentError),

    #[error("{0}")]
    Admin(#[from] AdminError),

    #[error("{0}")]
    Base(#[from] bs721_base::ContractError),

    #[error("NameNotFound")]
    NameNotFound {},

    #[error("AddressAlreadyMapped")]
    AddressAlreadyMapped {},

    #[error("RecordNameAlreadyExists")]
    RecordNameAlreadyExists {},

    #[error("RecordNameEmpty")]
    RecordNameEmpty {},

    #[error("RecordNameTooLong")]
    RecordNameTooLong {},

    #[error("RecordValueTooLong")]
    RecordValueTooLong {},

    #[error("RecordValueEmpty")]
    RecordValueEmpty {},

    #[error("UnauthorizedVerification")]
    UnauthorizedVerification {},

    #[error("Invalid Metadata")]
    InvalidMetadata {},

    #[error("Unauthorized: Not contract creator or admin")]
    UnauthorizedCreatorOrAdmin {},

    #[error("TooManyRecords max: {max}")]
    TooManyRecords { max: u32 },

    #[error("NotImplemented")]
    NotImplemented {},
}
