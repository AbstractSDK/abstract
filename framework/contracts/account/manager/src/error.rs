use abstract_core::{
    objects::{validation::ValidationError, version_control::VersionControlError},
    AbstractError,
};
use abstract_sdk::{core::objects::module::ModuleInfo, AbstractSdkError};
use cosmwasm_std::{Instantiate2AddressError, StdError};
use cw_controllers::AdminError;
use thiserror::Error;

#[derive(Error, Debug, PartialEq)]
pub enum ManagerError {
    #[error("{0}")]
    Std(#[from] StdError),

    #[error("{0}")]
    Abstract(#[from] AbstractError),

    #[error("{0}")]
    AbstractSdk(#[from] AbstractSdkError),

    #[error("{0}")]
    Admin(#[from] AdminError),

    #[error("{0}")]
    Validation(#[from] ValidationError),

    #[error("{0}")]
    Ownership(#[from] cw_ownable::OwnershipError),

    #[error("{0}")]
    Instantiate2AddressError(#[from] Instantiate2AddressError),

    #[error("{0}")]
    VersionControlError(#[from] VersionControlError),

    #[error("Module with id: {0} is already installed")]
    ModuleAlreadyInstalled(String),

    #[error("Cannot remove module because {0:?} depend(s) on it.")]
    ModuleHasDependents(Vec<String>),

    #[error("Contract got an unexpected Reply")]
    UnexpectedReply(),

    #[error("The name of the proposed module can not have length 0.")]
    InvalidModuleName {},

    #[error("Registering module fails because caller is not module factory")]
    CallerNotModuleFactory {},

    #[error("A migrate msg is required when when migrating this module")]
    MsgRequired {},

    #[error("{0} not upgradable")]
    NotUpgradeable(ModuleInfo),

    #[error("Cannot migrate {} twice", module_id)]
    DuplicateModuleMigration { module_id: String },

    #[error("Your account is currently suspended")]
    AccountSuspended {},

    #[error("The provided contract version {0} is lower than the current version {1}")]
    OlderVersion(String, String),

    #[error("The provided module {0} was not found")]
    ModuleNotFound(String),

    #[error("Module {module_id} with version {version} does not fit requirement {comp}, post_migration: {post_migration}")]
    VersionRequirementNotMet {
        module_id: String,
        version: String,
        comp: String,
        post_migration: bool,
    },

    #[error("module {0} is a dependency of {1} and is not installed.")]
    DependencyNotMet(String, String),

    #[error("The provided module {0} has an invalid module reference.")]
    InvalidReference(ModuleInfo),

    #[error("Cannot remove proxy")]
    CannotRemoveProxy {},

    #[error("No updates were included")]
    NoUpdates {},

    #[error("invalid configuration action, {}", error)]
    InvalidConfigAction { error: StdError },

    #[error("Must use ProposeOwner to change owner")]
    MustUseProposeOwner {},

    #[error("The address {0} doesn't have an owner, the manager can't determine admin right")]
    NoContractOwner(String),

    #[error("
            Checking the admin recursively failed. 
            You either have the an error in your sub-account configuration or you are not authorized to make this call.
    ")]
    SubAccountAdminVerification,

    #[error("Removing sub account failed")]
    SubAccountRemovalFailed {},

    #[error("Register of sub account failed")]
    SubAccountRegisterFailed {},

    #[error("Can't renounce account, with active sub account")]
    RenounceWithSubAccount {},

    #[error("Can't propose Renounced governance, use update_ownership instead")]
    ProposeRenounced {},

    #[error("Can't create account with Renounced governance")]
    InitRenounced {},
}
