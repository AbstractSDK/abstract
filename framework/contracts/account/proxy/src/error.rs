use abstract_core::{objects::ans_host::AnsHostError, AbstractError};
use abstract_sdk::AbstractSdkError;
use cosmwasm_std::{StdError, Uint128};
use cw_asset::AssetError;
use cw_utils::ParseReplyError;
use thiserror::Error;

#[derive(Error, Debug, PartialEq)]
pub enum ProxyError {
    #[error("{0}")]
    Std(#[from] StdError),

    #[error("{0}")]
    Abstract(#[from] AbstractError),

    #[error("{0}")]
    AbstractSdk(#[from] AbstractSdkError),

    #[error("{0}")]
    Asset(#[from] AssetError),

    #[error(transparent)]
    Admin(#[from] ::cw_controllers::AdminError),

    #[error("{0}")]
    Parse(#[from] ParseReplyError),

    #[error("{0}")]
    AnsHostError(#[from] AnsHostError),

    #[error("Module with address {0} is already whitelisted")]
    AlreadyWhitelisted(String),

    #[error("Module with address {0} not found in whitelist")]
    NotWhitelisted(String),

    #[error("Sender is not whitelisted")]
    SenderNotWhitelisted {},

    #[error("Max amount of assets registered")]
    AssetsLimitReached,

    #[error("Max amount of modules registered")]
    ModuleLimitReached,

    #[error("no base asset registered on proxy")]
    MissingBaseAsset,

    #[error("The proposed update resulted in a bad configuration: {0}")]
    BadUpdate(String),

    #[error(
        "Account balance too low, {} requested but it only has {}",
        requested,
        balance
    )]
    Broke {
        balance: Uint128,
        requested: Uint128,
    },

    #[error("Contract got an unexpected Reply")]
    UnexpectedReply(),
}
