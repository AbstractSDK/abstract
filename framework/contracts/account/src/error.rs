use abstract_sdk::std::objects::module::ModuleInfo;
use abstract_std::{
    objects::{registry::RegistryError, validation::ValidationError},
    AbstractError,
};
use cosmwasm_std::{Instantiate2AddressError, StdError};
use thiserror::Error;

#[derive(Error, Debug, PartialEq)]
pub enum AccountError {
    #[error(transparent)]
    Std(#[from] StdError),

    #[error(transparent)]
    Abstract(#[from] AbstractError),

    #[error(transparent)]
    Validation(#[from] ValidationError),

    #[error(transparent)]
    Ownership(#[from] abstract_std::objects::ownership::GovOwnershipError),

    #[error(transparent)]
    Instantiate2AddressError(#[from] Instantiate2AddressError),

    #[error(transparent)]
    RegistryError(#[from] RegistryError),

    #[error("Your account is currently suspended")]
    AccountSuspended {},

    // ** Modules Error ** //
    #[error("Failed to query modules to install: {error}")]
    QueryModulesFailed { error: RegistryError },

    #[error("Module with id: {0} is already installed")]
    ModuleAlreadyInstalled(String),

    #[error("Reinstalls of same version of app or standalone are not allowed")]
    ProhibitedReinstall {},

    #[error("The provided module {0} can't be installed on an Abstract account")]
    ModuleNotInstallable(String),

    #[error("The name of the proposed module can not have length 0.")]
    InvalidModuleName {},

    #[error("The provided module {0} was not found")]
    ModuleNotFound(String),

    #[error("Cannot migrate {} twice", module_id)]
    DuplicateModuleMigration { module_id: String },

    #[error("{0} not upgradable")]
    NotUpgradeable(ModuleInfo),

    #[error("Cannot remove module because {0:?} depend(s) on it.")]
    ModuleHasDependents(Vec<String>),

    #[error("Module {module_id} with version {version} does not fit requirement {comp}, post_migration: {post_migration}")]
    VersionRequirementNotMet {
        module_id: String,
        version: String,
        comp: String,
        post_migration: bool,
    },

    #[error("module {0} is a dependency of {1} and is not installed.")]
    DependencyNotMet(String, String),

    #[error("Max amount of modules registered")]
    ModuleLimitReached {},

    #[error("Module with address {0} is already whitelisted")]
    AlreadyWhitelisted(String),

    #[error("can't remove module that is not whitelisted")]
    NotWhitelisted {},

    // ** Sub Account ** //
    #[error("Removing sub account failed")]
    SubAccountRemovalFailed {},

    #[error("Register of sub account failed")]
    SubAccountRegisterFailed {},

    #[error("Can't renounce account, with active sub account")]
    RenounceWithSubAccount {},

    // ** Other Errors ** //
    #[error("No updates were included")]
    NoUpdates {},

    #[error("invalid configuration action, {}", error)]
    InvalidConfigAction { error: StdError },

    #[error("The provided contract version {0} is lower than the current version {1}")]
    OlderVersion(String, String),

    #[error("Sender is not whitelisted and is not a valid owner")]
    SenderNotWhitelistedOrOwner {},

    #[error("Contract got an unexpected Reply")]
    UnexpectedReply(),

    #[error("The caller ({caller}) is not the owner account's account ({account}). Only account can create sub-accounts for itself.", )]
    SubAccountCreatorNotAccount { caller: String, account: String },

    #[error("You can't chain admin calls")]
    CantChainAdminCalls {},

    #[error("Abstract Account Address don't match to the Contract address")]
    AbsAccInvalidAddr {
        abstract_account: String,
        contract: String,
    },

    #[error("Abstract Account don't have Authentication")]
    AbsAccNoAuth {},

    #[cfg(feature = "xion")]
    #[error(transparent)]
    AbstractXion(#[from] abstract_xion::error::ContractError),
}
