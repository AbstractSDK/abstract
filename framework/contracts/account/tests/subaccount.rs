use abstract_account::error::AccountError;
use abstract_integration_tests::{create_default_account, AResult};
use abstract_interface::*;
use abstract_std::{
    account::SubAccountIdsResponse,
    objects::{
        gov_type::{GovAction, GovernanceDetails},
        ownership,
    },
};
use cosmwasm_std::{to_json_binary, wasm_execute, WasmMsg};
use cw_orch::prelude::*;

#[test]
fn creating_on_subaccount_should_succeed() -> AResult {
    let chain = MockBech32::new("mock");
    let sender = chain.sender_addr();
    let deployment = Abstract::deploy_on_mock(chain.clone())?;
    let account = create_default_account(&sender, &deployment)?;
    account.create_and_return_sub_account(
        AccountDetails {
            name: "My subaccount".to_string(),
            description: None,
            link: None,
            namespace: None,
            install_modules: vec![],
            account_id: None,
        },
        &[],
    )?;
    let sub_accounts = account.sub_account_ids(None, None)?;
    assert_eq!(
        sub_accounts,
        SubAccountIdsResponse {
            // only one sub-account and it should be account_id 2
            sub_accounts: vec![2]
        }
    );
    take_storage_snapshot!(chain, "creating_on_subaccount_should_succeed");
    Ok(())
}

#[test]
fn updating_on_subaccount_should_succeed() -> AResult {
    let chain = MockBech32::new("mock");
    let sender = chain.sender_addr();
    let deployment = Abstract::deploy_on_mock(chain.clone())?;
    let account = create_default_account(&sender, &deployment)?;
    account.create_and_return_sub_account(
        AccountDetails {
            name: "My subaccount".to_string(),
            description: None,
            link: None,
            namespace: None,
            install_modules: vec![],
            account_id: None,
        },
        &[],
    )?;

    let new_desc = "new desc";
    account.update_info(Some(new_desc.to_string()), None, None)?;

    assert_eq!(Some(new_desc.to_string()), account.info()?.info.description);
    take_storage_snapshot!(chain, "updating_on_subaccount_should_succeed");
    Ok(())
}

#[test]
fn proxy_updating_on_subaccount_should_succeed() -> AResult {
    let chain = MockBech32::new("mock");
    let sender = chain.sender_addr();
    let deployment = Abstract::deploy_on_mock(chain.clone())?;
    let account = create_default_account(&sender, &deployment)?;
    let proxy_address = account.address()?;
    let sub_account = account.create_and_return_sub_account(
        AccountDetails {
            name: "My subaccount".to_string(),
            ..Default::default()
        },
        &[],
    )?;
    let new_desc = "new desc";

    // We call as the proxy, it should also be possible
    sub_account
        .call_as(&proxy_address)
        .update_info(Some(new_desc.to_owned()), None, None)?;

    assert_eq!(
        Some(new_desc.to_string()),
        sub_account.info()?.info.description
    );

    take_storage_snapshot!(chain, "proxy_updating_on_subaccount_should_succeed");
    Ok(())
}

#[test]
fn recursive_updating_on_subaccount_should_succeed() -> AResult {
    let chain = MockBech32::new("mock");
    let sender = chain.sender_addr();
    let deployment = Abstract::deploy_on_mock(chain.clone())?;
    let account = create_default_account(&sender, &deployment)?;
    let sub_account = account.create_and_return_sub_account(
        AccountDetails {
            name: "My subaccount".to_string(),
            ..Default::default()
        },
        &[],
    )?;
    // We call as the manager, it should also be possible
    let sub_sub_account = sub_account.create_and_return_sub_account(
        AccountDetails {
            name: "My subaccount".to_string(),
            ..Default::default()
        },
        &[],
    )?;
    let new_desc = "new desc";

    sub_sub_account
        .call_as(&sender)
        .update_info(Some(new_desc.to_string()), None, None)?;

    assert_eq!(
        Some(new_desc.to_string()),
        sub_sub_account.info()?.info.description
    );

    take_storage_snapshot!(chain, "recursive_updating_on_subaccount_should_succeed");
    Ok(())
}

#[test]
fn installed_app_updating_on_subaccount_should_succeed() -> AResult {
    let chain = MockBech32::new("mock");
    let sender = chain.sender_addr();
    let deployment = Abstract::deploy_on_mock(chain.clone())?;
    let account = create_default_account(&sender, &deployment)?;
    let sub_account = account.create_and_return_sub_account(
        AccountDetails {
            name: "My subaccount".to_string(),
            ..Default::default()
        },
        &[],
    )?;

    let mock_app = chain.addr_make("mock_app");
    account.update_whitelist(vec![mock_app.to_string()], Vec::default())?;

    let new_desc = "new desc";
    // adding mock_app to whitelist on proxy

    // We call as installed app of the owner-proxy, it should also be possible
    account.call_as(&mock_app).execute_msgs(
        vec![wasm_execute(
            sub_account.addr_str()?,
            &abstract_std::account::ExecuteMsg::<Empty>::UpdateInfo {
                name: None,
                description: Some(new_desc.to_owned()),
                link: None,
            },
            vec![],
        )?
        .into()],
        &[],
    )?;

    assert_eq!(
        Some(new_desc.to_string()),
        sub_account.info()?.info.description
    );
    take_storage_snapshot!(chain, "installed_app_updating_on_subaccount_should_succeed");

    Ok(())
}

#[test]
fn sub_account_move_ownership() -> AResult {
    let chain = MockBech32::new("mock");
    let sender = chain.sender_addr();
    let new_owner = chain.addr_make("new_owner");
    let deployment = Abstract::deploy_on_mock(chain.clone())?;
    let account = create_default_account(&sender, &deployment)?;
    // Store manager address, it will be used for querying
    let manager_addr = account.address()?;

    let sub_account = account.create_and_return_sub_account(
        AccountDetails {
            name: "My subaccount".to_string(),
            ..Default::default()
        },
        &[],
    )?;

    let sub_accounts = account.sub_account_ids(None, None)?;
    assert_eq!(
        sub_accounts,
        SubAccountIdsResponse {
            // only one sub-account and it should be account_id 2
            sub_accounts: vec![2]
        }
    );

    sub_account.update_ownership(GovAction::TransferOwnership {
        new_owner: GovernanceDetails::Monarchy {
            monarch: new_owner.to_string(),
        },
        expiry: None,
    })?;

    // Make sure it's not updated until claimed
    let sub_accounts: SubAccountIdsResponse = chain.query(
        &abstract_std::account::QueryMsg::SubAccountIds {
            start_after: None,
            limit: None,
        },
        &manager_addr,
    )?;
    assert_eq!(
        sub_accounts,
        SubAccountIdsResponse {
            sub_accounts: vec![2]
        }
    );

    // Claim ownership
    sub_account.call_as(&new_owner).execute(
        &abstract_std::account::ExecuteMsg::UpdateOwnership(ownership::GovAction::AcceptOwnership),
        &[],
    )?;

    // After claim it's updated
    let sub_accounts = account.sub_account_ids(None, None)?;
    assert_eq!(
        sub_accounts,
        SubAccountIdsResponse {
            sub_accounts: vec![]
        }
    );
    take_storage_snapshot!(chain, "sub_account_move_ownership");

    Ok(())
}

#[test]
fn account_move_ownership_to_sub_account() -> AResult {
    let chain = MockBech32::new("mock");
    Abstract::deploy_on_mock(chain.clone())?;
    abstract_integration_tests::account::account_move_ownership_to_sub_account(chain)?;
    Ok(())
}

#[test]
fn sub_account_move_ownership_to_sub_account() -> AResult {
    let chain = MockBech32::new("mock");
    let sender = chain.sender_addr();
    let deployment = Abstract::deploy_on_mock(chain.clone())?;
    let account = create_default_account(&sender, &deployment)?;

    let sub_account = account.create_and_return_sub_account(
        AccountDetails {
            name: "My subaccount".to_string(),
            ..Default::default()
        },
        &[],
    )?;
    let sub_account_addr = sub_account.address()?;

    let new_account = create_default_account(&sender, &deployment)?;

    let new_account_sub_account = new_account.create_and_return_sub_account(
        AccountDetails {
            name: "My subaccount".to_string(),
            ..Default::default()
        },
        &[],
    )?;

    // sub-accounts state updated
    let sub_ids = new_account.sub_account_ids(None, None)?;
    assert_eq!(sub_ids.sub_accounts, vec![4]);

    let new_governance = GovernanceDetails::SubAccount {
        account: sub_account_addr.to_string(),
    };
    new_account_sub_account.update_ownership(GovAction::TransferOwnership {
        new_owner: new_governance.clone(),
        expiry: None,
    })?;
    let new_account_sub_account_addr = new_account_sub_account.address()?;

    let mock_module = chain.addr_make("mock_module");

    // Should error as the ownership is not accepted yet
    new_account_sub_account
        .call_as(&sub_account_addr)
        .update_whitelist(vec![mock_module.to_string()], Vec::default())
        .expect_err("ownership not accepted yet.");

    sub_account.execute_msgs(
        vec![wasm_execute(
            new_account_sub_account_addr,
            &abstract_std::account::ExecuteMsg::<Empty>::UpdateOwnership(
                ownership::GovAction::AcceptOwnership,
            ),
            vec![],
        )?
        .into()],
        &[],
    )?;

    // sub-accounts state updated
    let sub_ids = sub_account.sub_account_ids(None, None)?;
    assert_eq!(sub_ids.sub_accounts, vec![4]);
    // removed from the previous owner as well
    let sub_ids = new_account.sub_account_ids(None, None)?;
    assert_eq!(sub_ids.sub_accounts, Vec::<u32>::new());

    new_account_sub_account
        .call_as(&sub_account_addr)
        .update_whitelist(vec![mock_module.to_string()], Vec::default())?;

    new_account_sub_account.expect_whitelist(vec![mock_module])?;

    let info = new_account_sub_account.ownership()?;
    assert_eq!(new_governance, info.owner);
    take_storage_snapshot!(chain, "sub_account_move_ownership_to_sub_account");

    Ok(())
}

#[test]
fn account_updated_to_subaccount() -> AResult {
    let chain = MockBech32::new("mock");
    let sender = chain.sender_addr();
    let deployment = Abstract::deploy_on_mock(chain.clone())?;

    // Creating account1
    let account_1 = create_default_account(&sender, &deployment)?;

    // Creating account2
    let account_2 = create_default_account(&sender, &deployment)?;

    // Setting account1 as pending owner of account2
    account_2.update_ownership(GovAction::TransferOwnership {
        new_owner: GovernanceDetails::SubAccount {
            account: account_1.addr_str()?,
        },
        expiry: None,
    })?;

    // account1 accepting account2 as a sub-account
    let accept_msg = abstract_std::account::ExecuteMsg::<Empty>::UpdateOwnership(
        ownership::GovAction::AcceptOwnership,
    );
    account_1.execute_msgs(
        vec![wasm_execute(account_2.addr_str()?, &accept_msg, vec![])?.into()],
        &[],
    )?;

    // Check account_1 knows about his new sub-account
    let ids = account_1.sub_account_ids(None, None)?;
    assert_eq!(ids.sub_accounts.len(), 1);
    Ok(())
}

#[test]
fn account_updated_to_subaccount_recursive() -> AResult {
    let chain = MockBech32::new("mock");
    let sender = chain.sender_addr();
    let deployment = Abstract::deploy_on_mock(chain.clone())?;

    // Creating account1
    let account_1 = create_default_account(&sender, &deployment)?;

    // Creating account2
    let account_2 = create_default_account(&sender, &deployment)?;

    // Setting account1 as pending owner of account2
    account_2.update_ownership(GovAction::TransferOwnership {
        new_owner: GovernanceDetails::SubAccount {
            account: account_1.addr_str()?,
        },
        expiry: None,
    })?;
    // accepting ownership by sender instead of the manager
    account_2.update_ownership(ownership::GovAction::AcceptOwnership)?;

    // Check manager knows about his new sub-account
    let ids = account_1.sub_account_ids(None, None)?;
    assert_eq!(ids.sub_accounts.len(), 1);
    Ok(())
}

#[test]
fn top_level_owner() -> AResult {
    let chain = MockBech32::new("mock");
    let sender = chain.sender_addr();
    let deployment = Abstract::deploy_on_mock(chain.clone())?;

    let account = create_default_account(&sender, &deployment)?;
    let sub_account = account.create_and_return_sub_account(
        AccountDetails {
            name: "My subaccount".to_string(),
            ..Default::default()
        },
        &[],
    )?;

    let top_level_owner = sub_account.top_level_owner()?;
    assert_eq!(top_level_owner.address, sender);
    Ok(())
}

#[test]
fn cant_renounce_with_sub_accounts() -> AResult {
    let chain = MockBech32::new("mock");
    let sender = chain.sender_addr();
    let deployment = Abstract::deploy_on_mock(chain.clone())?;

    let account = create_default_account(&sender, &deployment)?;
    // Creating sub account
    account.create_and_return_sub_account(
        AccountDetails {
            name: "My subaccount".to_string(),
            ..Default::default()
        },
        &[],
    )?;

    let err: AccountError = account
        .update_ownership(ownership::GovAction::RenounceOwnership)
        .unwrap_err()
        .downcast()
        .unwrap();
    assert_eq!(err, AccountError::RenounceWithSubAccount {});
    Ok(())
}

#[test]
fn can_renounce_sub_accounts() -> AResult {
    let chain = MockBech32::new("mock");
    let sender = chain.sender_addr();
    let deployment = Abstract::deploy_on_mock(chain.clone())?;

    let account = create_default_account(&sender, &deployment)?;
    // Creating sub account
    let sub_account = account.create_and_return_sub_account(
        AccountDetails {
            name: "My subaccount".to_string(),
            ..Default::default()
        },
        &[],
    )?;

    sub_account.update_ownership(ownership::GovAction::RenounceOwnership)?;

    account.update_ownership(ownership::GovAction::RenounceOwnership)?;

    // No owners
    // Renounced governance
    let account_owner = account.ownership()?;
    assert_eq!(account_owner.owner, GovernanceDetails::Renounced {});
    let sub_account_owner = sub_account.ownership()?;
    assert_eq!(sub_account_owner.owner, GovernanceDetails::Renounced {});

    Ok(())
}

#[test]
fn account_updated_to_subaccount_without_recursion() -> AResult {
    let chain = MockBech32::new("mock");
    let sender = chain.sender_addr();
    let deployment = Abstract::deploy_on_mock(chain.clone())?;

    // Creating account1
    let account_1 = create_default_account(&sender, &deployment)?;

    // Creating account2
    let account_2 = create_default_account(&sender, &deployment)?;

    // Setting account1 as pending owner of account2
    account_2.update_ownership(GovAction::TransferOwnership {
        new_owner: GovernanceDetails::SubAccount {
            account: account_1.addr_str()?,
        },
        expiry: None,
    })?;

    // accepting ownership by sender instead of the manager
    account_1.execute_msgs(
        vec![WasmMsg::Execute {
            contract_addr: account_2.addr_str()?,
            msg: to_json_binary(
                &abstract_std::account::ExecuteMsg::<Empty>::UpdateOwnership(
                    GovAction::AcceptOwnership,
                ),
            )?,
            funds: Vec::default(),
        }
        .into()],
        &[],
    )?;

    // Check manager knows about his new sub-account
    let ids = account_1.sub_account_ids(None, None)?;
    assert_eq!(ids.sub_accounts.len(), 1);
    Ok(())
}

#[test]
fn sub_account_to_regular_account_without_recursion() -> AResult {
    let chain = MockBech32::new("mock");
    let sender = chain.sender_addr();
    let deployment = Abstract::deploy_on_mock(chain.clone())?;

    // Creating account1
    let account = create_default_account(&sender, &deployment)?;
    let sub_account = account.create_and_return_sub_account(
        AccountDetails {
            name: "My subaccount".to_string(),
            ..Default::default()
        },
        &[],
    )?;

    account.execute_msgs(
        vec![WasmMsg::Execute {
            contract_addr: sub_account.addr_str()?,
            msg: to_json_binary(&abstract_account::msg::ExecuteMsg::UpdateOwnership(
                GovAction::TransferOwnership {
                    new_owner: GovernanceDetails::Monarchy {
                        monarch: chain.sender_addr().to_string(),
                    },
                    expiry: None,
                },
            ))?,
            funds: vec![],
        }
        .into()],
        &[],
    )?;

    sub_account.update_ownership(GovAction::AcceptOwnership)?;
    let ownership = sub_account.ownership()?;
    assert_eq!(
        ownership.owner,
        GovernanceDetails::Monarchy {
            monarch: chain.sender_addr().to_string()
        }
    );
    Ok(())
}
