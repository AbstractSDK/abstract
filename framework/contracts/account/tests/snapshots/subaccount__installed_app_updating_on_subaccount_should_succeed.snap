---
source: contracts/account/tests/subaccount.rs
expression: all_storage
---
"abstract:account-local-0":
  - - aa
    - "false"
  - - ab
    - "{\"name\":\"Default Abstract Account\"}"
  - - af
    - "[]"
  - - ag
    - "{\"trace\":\"local\",\"seq\":0}"
  - - ai
    - "[]"
  - - contract_info
    - "{\"contract\":\"abstract:account\",\"version\":\"0.24.1-beta.1\"}"
  - - ownership
    - "{\"owner\":{\"monarchy\":{\"monarch\":\"mock14cl2dthqamgucg9sfvv4relp3aa83e40yfzzc8\"}},\"pending_owner\":null,\"pending_expiry\":null}"
"abstract:account-local-1":
  - - "\u0000\u0002ae\u0000\u0000\u0000\u0002"
    - "{}"
  - - aa
    - "false"
  - - ab
    - "{\"name\":\"Default Abstract Account\"}"
  - - af
    - "[\"mock15cnh8a268g6dhlzc4cay9mq7w8vt24hmknafxpngvmev4t7dj0jq607wev\"]"
  - - ag
    - "{\"trace\":\"local\",\"seq\":1}"
  - - ai
    - "[]"
  - - contract_info
    - "{\"contract\":\"abstract:account\",\"version\":\"0.24.1-beta.1\"}"
  - - ownership
    - "{\"owner\":{\"monarchy\":{\"monarch\":\"mock1pgm8hyk0pvphmlvfjc8wsvk4daluz5tgrw6pu5mfpemk74uxnx9qwrtv4f\"}},\"pending_owner\":null,\"pending_expiry\":null}"
"abstract:account-local-2":
  - - aa
    - "false"
  - - ab
    - "{\"name\":\"My subaccount\",\"description\":\"new desc\"}"
  - - af
    - "[]"
  - - ag
    - "{\"trace\":\"local\",\"seq\":2}"
  - - ai
    - "[]"
  - - contract_info
    - "{\"contract\":\"abstract:account\",\"version\":\"0.24.1-beta.1\"}"
  - - ownership
    - "{\"owner\":{\"sub_account\":{\"account\":\"mock1jdl4tasw0x00k2edyf8656lqxy95gx06h2nt5n8qjzv45yxr5ekq0myk0q\"}},\"pending_owner\":null,\"pending_expiry\":null}"
"abstract:ans-host":
  - - be
    - "[]"
  - - cfg
    - "{\"next_unique_pool_id\":1}"
  - - contract_info
    - "{\"contract\":\"abstract:ans-host\",\"version\":\"0.24.1-beta.1\"}"
  - - ownership
    - "{\"owner\":\"mock14cl2dthqamgucg9sfvv4relp3aa83e40yfzzc8\",\"pending_owner\":null,\"pending_expiry\":null}"
"abstract:ibc-client":
  - - contract_info
    - "{\"contract\":\"abstract:ibc-client\",\"version\":\"0.24.1-beta.1\"}"
  - - mod
    - "{\"module\":\"abstract:ibc-client\",\"version\":\"0.24.1-beta.1\",\"dependencies\":[],\"metadata\":null}"
  - - ownership
    - "{\"owner\":\"mock14cl2dthqamgucg9sfvv4relp3aa83e40yfzzc8\",\"pending_owner\":null,\"pending_expiry\":null}"
"abstract:ibc-host":
  - - contract_info
    - "{\"contract\":\"abstract:ibc-host\",\"version\":\"0.24.1-beta.1\"}"
  - - ownership
    - "{\"owner\":\"mock14cl2dthqamgucg9sfvv4relp3aa83e40yfzzc8\",\"pending_owner\":null,\"pending_expiry\":null}"
"abstract:module-factory":
  - - contract_info
    - "{\"contract\":\"abstract:module-factory\",\"version\":\"0.24.1-beta.1\"}"
  - - ownership
    - "{\"owner\":\"mock14cl2dthqamgucg9sfvv4relp3aa83e40yfzzc8\",\"pending_owner\":null,\"pending_expiry\":null}"
"abstract:registry":
  - - "\u0000\u0002cb\u0000\babstract\u0000\u0007account0.24.1-beta.1"
    - "{\"account\":5}"
  - - "\u0000\u0002cb\u0000\babstract\u0000\bans-host0.24.1-beta.1"
    - "{\"native\":\"mock1wx0qjtlz799pfxl73y2f4dv28nukztpmq2ztavk60v534tat9cds0pnytz\"}"
  - - "\u0000\u0002cb\u0000\babstract\u0000\bibc-host0.24.1-beta.1"
    - "{\"native\":\"mock1ldaf3e9zxja5mrcw3clcu873wggmqhwfvqw9djjcjxxjwv59crqsc6n7r3\"}"
  - - "\u0000\u0002cb\u0000\babstract\u0000\bregistry0.24.1-beta.1"
    - "{\"native\":\"mock1cag6cwygef6fuddgq2l44py7crpscufrza4mt3rum2axueemenmqapxshc\"}"
  - - "\u0000\u0002cb\u0000\babstract\u0000\nibc-client0.24.1-beta.1"
    - "{\"native\":\"mock1g86ckugm7v9vvpadq9szg6rlf72pdl4a6n0d8apltq5t0grecpwqtrcaj0\"}"
  - - "\u0000\u0002cb\u0000\babstract\u0000\u000emodule-factory0.24.1-beta.1"
    - "{\"native\":\"mock1vnj802lyyupfaez73w7axfs3xveraxen370vhcnumaufj29r5rxqwxdk4n\"}"
  - - "\u0000\u0002ch\u0000\u0005local\u0000\u0000\u0000\u0000"
    - "\"mock1ys8clus4qf2haemyck0uljz74dj2k8hu4ggfpe7vty6zd5j2982q9sf90k\""
  - - "\u0000\u0002ch\u0000\u0005local\u0000\u0000\u0000\u0001"
    - "\"mock1jdl4tasw0x00k2edyf8656lqxy95gx06h2nt5n8qjzv45yxr5ekq0myk0q\""
  - - "\u0000\u0002ch\u0000\u0005local\u0000\u0000\u0000\u0002"
    - "\"mock1zrcvn33rptsj7t6hacjrhuekqplf8eagqs3dg7m9vqqrhnyaxtxsa9tlkx\""
  - - "\u0000\u0002cjabstract"
    - "{\"trace\":\"local\",\"seq\":0}"
  - - cfg
    - "{\"security_disabled\":true,\"namespace_registration_fee\":null}"
  - - ci
    - "3"
  - - contract_info
    - "{\"contract\":\"abstract:registry\",\"version\":\"0.24.1-beta.1\"}"
  - - ownership
    - "{\"owner\":\"mock14cl2dthqamgucg9sfvv4relp3aa83e40yfzzc8\",\"pending_owner\":null,\"pending_expiry\":null}"
